### ISSUE TITLE
Support role="none" on <li> elements inside role="menu"

### ISSUE THREAD

**Issue by ma-5** [validator/validator#1970]
### Spec URL

https://dequeuniversity.com/rules/axe/4.11/aria-allowed-role | https://html.spec.whatwg.org/multipage/dom.html#allowed-aria-roles,-states-and-properties

### Test case

```html
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Page Title</title>
</head>
<body>
<ul role="menu">
  <li role="none">
    <button role="menuitem">Start</button>
  </li>
</ul>
<div>
<p>Currently, the Nu HTML Checker reports the following error for the snippet above:</p>

<p>Error: An li element that is a descendant of a role=menu element or role=menubar element must not have any role value other than group, menuitem, menuitemcheckbox, menuitemradio, or separator.</p>

<p>However, according to accessibility guidance (see Deque University link above), role="none" is a valid role
for &lt;li&gt; elements in this context. This allows authors to suppress the implicit list semantics while 
providing explicit menuitem roles on interactive descendants (e.g., &lt;button role="menuitem"&gt;).</p>

<p>I would like to request that the validator be updated to accept role="none" on &lt;li&gt; elements inside role="menu" or role="menubar", 
in line with current accessibility best practices and ARIA specifications.</p>
</div>
</body>
</html>
```

**Comment by sideshowbarker** [comment]
@scottaohara @pkra Can you confirm if the checker behavior here conforms to the current ARIA in HTML requirements?

This message:

> _Error: An `li` element that is a descendant of a `role=menu` element or `role=menubar` element must not have any role value other than `group`, `menuitem`, `menuitemcheckbox`, `menuitemradio`, or `separator`._

…(cited in the issue description) is a new message the now gets reported due to the changes I made for https://github.com/validator/validator/issues/1368.

**Comment by pkra** [comment]
@scottaohara will know better but the example seem valid to me. 

Per https://w3c.github.io/html-aria/#el-li an `li` can have

> any role if the parent list element does not expose an implicit or explicit list role. 

The `menu` role is not a list role (and is allowed on `ul`).

**Comment by sideshowbarker** [comment]
> Per [w3c.github.io/html-aria#el-li](https://w3c.github.io/html-aria/#el-li) an `li` can have
> 
> > any role if the parent list element does not expose an implicit or explicit list role.

OK then I guess my next question would be: Is that statement in the ARIA in HTML spec actually correct per the requirements in the ARIA spec?

I ask that because, at https://w3c.github.io/aria/#mustContain in the ARIA spec, I see this requirement:

> _Authors MUST only add child element with allowed roles. For example, an element with the role [list](https://w3c.github.io/aria/#list) can own child elements with the role [listitem](https://w3c.github.io/aria/#listitem), but cannot own elements with the role [option](https://w3c.github.io/aria/#option)._

And to determine what the “allowed roles” are for a child element of a `role="menu"` element, I look at the _“Allowed Accessibility Child Roles”_ row of the _“Characteristics”_ table for the `menu` role at https://w3c.github.io/aria/#menu — and I find that says the allowed roles are:

> - [group](https://w3c.github.io/aria/#group) with [accessibility child](https://w3c.github.io/aria/#dfn-accessibility-child) [menuitem](https://w3c.github.io/aria/#menuitem)
> - [group](https://w3c.github.io/aria/#group) with [accessibility child](https://w3c.github.io/aria/#dfn-accessibility-child) [menuitemradio](https://w3c.github.io/aria/#menuitemradio)
> - [group](https://w3c.github.io/aria/#group) with [accessibility child](https://w3c.github.io/aria/#dfn-accessibility-child) [menuitemcheckbox](https://w3c.github.io/aria/#menuitemcheckbox)
> - [menuitem](https://w3c.github.io/aria/#menuitem)
> - [menuitemcheckbox](https://w3c.github.io/aria/#menuitemcheckbox)
> - [menuitemradio](https://w3c.github.io/aria/#menuitemradio)
> - [separator](https://w3c.github.io/aria/#separator)

…which doesn’t include the `none` role. And so, per those ARIA spec requirements at least, that `li` child of that `role="menu"` `ul` element isn’t allowed to have `role=none`.

Or, is there something I’m misunderstanding?

**Comment by pkra** [comment]
Right. I think you're missing https://w3c.github.io/aria/#mustContain

> To determine whether an element is the child of an element, user agents MUST ignore any intervening elements with the role generic or none.

**Comment by sideshowbarker** [comment]
> Right. I think you're missing [w3c.github.io/aria#mustContain](https://w3c.github.io/aria/#mustContain)
> 
> > To determine whether an element is the child of an element, user agents MUST ignore any intervening elements with the role generic or none.

OK, so to be clear, given this tree:

```html
<ul role="menu">
  <li role="none">
    <button role="menuitem">Start</button>
  </li>
</ul>
```

…then, do the following statements describe how the ARIA requirements are meant to be applied to that tree?

- That `li` element has _stopped_ being a child element of that `ul` element?
- And instead, that `li` element has become an “intervening element”?
- And the only way to determine if an `li` element like that _may_ have changed from being a child element and instead gotten transformed into being a “intervening element” is to:
   1.  Look _down the tree at the `li` element’s **descendants**._
   2. See if that `li` element has any descendants that happen to have one of the “Allowed Accessibility Child Roles” which the ARIA spec allows a `role="menu"` element to have?

And so, in other words, in the following case:

```html
<ul role="menu">
  <li role="none"></li>
</ul>
```
…the ARIA requirements _do_ consider that `<li role="none">` to be a child element of that `<ul role="menu">` element?

But in contrast, in the following case:

```html
<ul role="menu">
  <li role="none">
    <button role="menuitem">Start</button>
  </li>
</ul>
```
…the ARIA requirements _do not_ consider that `<li role="none">` to be a child element of that `<ul role="menu">` element?

And instead the _only_ “child element” is that `<button role="menuitem">`? (which, in every other context, literally everywhere else outside of the context of the ARIA requirements as currently written, is in no way at all considered a child element of that `ul` element at all — but instead, is a descendant of it.)

**Comment by pkra** [comment]
I'm not sure I understand these questions correctly but I would agree that the use of "child" is confusing.

In the quote I gave, "child" means "accessibility child" (cf. the paragraph preceding the quote).

Section 7.3 has the definition:

>  The **accessibility children** of a DOM element are all of the children of that element's corresponding accessible object in the accessibility tree.

While Section 7.1 says

>  The following elements are not exposed via the accessibility API and user agents MUST NOT include them in the accessibility tree: 
> [....]
> - Elements with none or presentation as the first role in the role attribute [...]  In addition, the element's descendants and text content are generally included.

So an element with role=none is never an "accessibility child" of anything because it's never in the accessibility tree. But its DOM children will generally still be in the accessibility tree. It's a bit like CSS's `display:contents`. 

Hope this helps. Otherwise, please file an issue with ARIA to clarify.

(FWIW these kinds of wordings used to be even messier before the accessibility tree was introduced into the spec. I'm sure there are more than a few warts like this in the spec.)

**Comment by sideshowbarker** [comment]
OK, please bear with me. I guess what I am asking for here is basically just what y’all want the checker behavior for to be _all_ the cases of an `li` element that has `role="none"`.

So, that’s what I’d like for somebody to clearly confirm here.

So, based on the examples in https://github.com/validator/validator/issues/1368, I assumed that issue was requesting that I add new error messages to the checker for the cases where an `li` has a role that’s not allowed by the tree the `li` is in.

And one example that was given there is this:

```html
<ul role=menu>
  <li role=button>...</li>
</ul>
```
So, from that, I have assumed what’s wanted for that is: the checker emits a message saying something like this:

> _Error: An `li` element that is a descendant of a `role=menu` element or `role=menubar` element must not have any role value other than `group`, `menuitem`, `menuitemcheckbox`, `menuitemradio`, or `separator`._

Am I correct about that least? I mean, is that what the issue is requesting?

And if so, then what is the issue asking that I do for the following case?

```html
<ul role="menu">
  <li role="none"></li>
</ul>
```
_Should_ the checker also emit the above error message for that case, too? Or should it _not_ for that case?

Based on what’s stated in https://github.com/validator/validator/issues/1970#issuecomment-3642866749, it seems like the checker should _not_ emit the error for that case — because _“an element with role=none is never an "accessibility child" of anything because it's never in the accessibility tree”_.

If that _is_ the case, then I can certainly update the checker to do that.

Specifically, I can update the checker to completely ignore every `li` element that has `role="none"` — and more specifically, I mean, I can make the checker stop emitting that _“An `li` element… must not have any role value other than…”_ message for any cases of an `li` element that has `role="none"`.

Is that what I should do?

**Comment by sideshowbarker** [comment]
> Otherwise, please file an issue with ARIA to clarify.

https://github.com/w3c/aria/pull/2697

**Comment by sideshowbarker** [comment]
> So an element with role=none is never an "accessibility child" of anything because it's never in the accessibility tree. But its DOM children will generally still be in the accessibility tree. It's a bit like CSS's `display:contents`.

Thanks — that explanation makes it clear to me. Raised https://github.com/w3c/aria/pull/2698 to get that added to the spec, as Note.

**Comment by pkra** [comment]
I'd say generic and none/presentation roles are allowed almost everywhere. 

However, there are exceptions. In ARIA itself, I think that's mostly from Section 9.3 (ok, technically, there's no role=none when that happens).

But host languages can also add constraints. In particular, (to touch on the second point): for HTML, an `li` with role=none is not allowed inside lists. 

So the checker cannot generally ignore all `li` with role=none; it will need to check if it's parent is an implicit or explicit list.

**Comment by sideshowbarker** [comment]
> So the checker cannot generally ignore all `li` with role=none; it will need to check if it's parent is an implicit or explicit list.

OK — that, I think, is do-able. I will try that.

(What would in contrast _not_ be quite a bit less readily do-able for the HTML checker would be: If for an `li`, I had to look _down_ the tree — lookahead — at the `li` element’s descendants. I had worried earlier that the spec requirements might necessitate doing that — so I’m relieved to find out that’s not actually required, at least.)

**Comment by sideshowbarker** [comment]
@ma-5 Thanks for having reported this. I’ve fixed it now, and deployed the fix:

https://validator.w3.org/nu/?showsource=yes&doc=data%3Atext%2Fhtml%3Bbase64%2CPCFkb2N0eXBlIGh0bWw%2BPGh0bWwgbGFuZz48bWV0YSBjaGFyc2V0PXV0Zi04Pgo8dGl0bGU%2BdGVzdDwvdGl0bGU%2BCjx1bCByb2xlPSJtZW51Ij4KICA8bGkgcm9sZT0ibm9uZSI%2BCiAgICA8YnV0dG9uIHJvbGU9Im1lbnVpdGVtIj5TdGFydDwvYnV0dG9uPgogIDwvbGk%2BCjwvdWw%2BCg%3D%3D
